sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3410,6 +3410,28 @@ case class ArrayDistinct(child: Expression)
     case _ => false
   }
 
+  @transient protected lazy val canUseSpecializedHashSet = elementType match {
```

Comment:
Can we extract those common methods?

Reply:
We can do so. To minimize the changes due to remaining time for cutting, I would like to do this in another PR #21912.